espresso/environment/2_espresso_liveness_test.go

```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+		)),
+	)
+
+	defer env.Stop(t, server)
+	system, espressoDevNode, err := launcher.StartDevNet(ctx, t, 0, option)
+
+	// Signal the testnet to shut down
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start dev environment with espresso dev node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	defer env.Stop(t, system)
+	defer env.Stop(t, espressoDevNode)
+
+	caffNode, err := env.LaunchDecaffNode(t, system, espressoDevNode)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start caff node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	// Shut down the Caff Node
+	defer env.Stop(t, caffNode)
+
+	addressAlice := system.Cfg.Secrets.Addresses().Alice
+
+	l1Client := system.NodeClient(e2esys.RoleL1)
+	l2Seq := system.NodeClient(e2esys.RoleSeq)
+	caffVerif := system.NodeClient(env.RoleCaffNode)
+
+	balanceAliceInitial, err := caffVerif.BalanceAt(ctx, addressAlice, nil)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("Failed to fetch Alice's balance:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	type espressoReceived struct {
+		batch    *derive.EspressoBatch
+		block    *geth_types.Block
+		received time.Time
+	}
+
+	espressoReceipts := map[common.Hash]espressoReceived{}
+
+	streamBlocksCtx, streamBlocksCancel := context.WithCancel(ctx)
+	var wg sync.WaitGroup
+	defer streamBlocksCancel()
+	{
+		// Streamer Setup and Configuration
+		l := log.NewLogger(slog.Default().Handler())
+		streamer := espresso.NewEspressoStreamer(
+			system.RollupConfig.L2ChainID.Uint64(),
+			batcher.NewAdaptL1BlockRefClient(l1Client),
+			espressoClient.NewClient(server.URL),
+			nil, // TODO(AG)
+			l,
+			func(b []byte) (*derive.EspressoBatch, error) {
+				return derive.UnmarshalEspressoTransaction(b, system.RollupConfig.Genesis.SystemConfig.BatcherAddr)
+			},
+			100*time.Millisecond,
+		)
+
+		l1Client, _ := client.NewRPC(streamBlocksCtx, l, system.NodeEndpoint(e2esys.RoleL1).RPC())
+		l2Seq, _ := client.NewRPC(streamBlocksCtx, l, system.NodeEndpoint(e2esys.RoleSeq).RPC())
+
+		l1RefClient, err := sources.NewL1Client(l1Client, l, nil, sources.L1ClientDefaultConfig(system.RollupConfig, true, sources.RPCKindStandard))
+		require.NoError(t, err, "failed to create L1 Ref client")
+		l2RefClient, err := sources.NewL2Client(l2Seq, l, nil, sources.L2ClientDefaultConfig(system.RollupConfig, true))
+		require.NoError(t, err, "failed to create L2 Ref client")
+		l2BlockRef, err := l2RefClient.L2BlockRefByLabel(streamBlocksCtx, eth.Safe)
+		require.NoError(t, err, "failed to get safe L2 block ref")
+		finalizedL1BlockRef, err := l1RefClient.L1BlockRefByLabel(streamBlocksCtx, eth.Finalized)
+		require.NoError(t, err, "failed to get finalized L1 block ref")
+		streamer.Refresh(streamBlocksCtx, finalizedL1BlockRef, l2BlockRef.Number)
+
+		// Start consuming Batches from the Streamer
+		// We cannot guarantee the that we will receive only the batches that
```

Comment:
```suggestion
		// We cannot guarantee that we will receive only the batches that
```

Reply:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.